### ISSUE TITLE
Make DIALS work with Python 3

### ISSUE THREAD

**Issue by graeme-winter** [dials/dials#843]
TL;DR - https://github.com/cctbx/dxtbx/issues/36 for DIALS

---------

# Python 3 for DIALS

As you are aware DIALS is already python 3 syntax compatible, and has been for some time. With cctbx and dxtbx now being nominally python 3 compatible we can proceed with our work on DIALS.

As discussed on dials-support we propose that this is done in a staged way, with open work and continuous testing to record progress. 

## Process

Before the work is started we will set up a Jenkins build job at Diamond and a python3 build job on Travis following the same principle that we had with the python 3 syntax conversion, ie. we will keep a list of expected failing tests in python 3 alongside the code.

Work will proceed in feature branches off master which will be submitted as pull requests, reviewed, and merged in a timely fashion i.e. over a period of a couple of days.
This ensures we do not have long-running side branches leading to code conflicts and the need to merge from the master branch or rebase the feature branch.
The entire test suite is run on the iteratively updated master branch which means we avoid python 2 regressions and - as the list of ignored tests shrinks - python 3 regressions.

Every feature branch will be concerned with a small number of topics. Initially these will be futurize stage 2 fixers, later these will be picked from remaining python 3 test failures. Commits in each pull request may fall into one of the following categories:
* automatic code conversion
* manual changes for test fixing commits
* idiomatic updates (eg. *enumerate()*, use of context managers, reliance on deprecated code, explicit list construction in places where generators would be more appropriate, ...)
* cleanup of immediate surrounding code (ie. cleaning campsite)
* improving code coverage, including adding tests
* flake8 compliance
* addressing of LGTM issues.

There is no required flake8/LGTM compliance level for pull requests, and particularly at the start of the process it may be the case that pull requests will be merged with failing flake8 tests. As we refactor DIALS and our code quality improves we should aim for more stringent flake8 compliance and start enforcing more flake8 test categories via the pre-commit hook. Pull requests should not introduce any new LGTM issues, and ideally address existing issues in the touched files, although this may again not be practicable at the start of the process.

## Benefits

At the end of the process DIALS will be idiomatically modernized python 3 code that is fully python 2.7 compatible. This will improve the overall code quality, reduce technical debt and code maintenance cost, and reduce the barriers to entry for new or external developers.

## Success criteria
* No test failures on Travis (Python 2.7 and 3.6)
* No test failures on DLS Jenkins (Python 2.7 and 3.6)
* Fewer than 30 remaining LGTM alerts (current: 213) with an A+ code rating (current: A+)
* Fewer than 30 flake8 errors/warnings (current: 1452)
* Delivery by the end of October 2019

## Comments

Please discuss this process in the issue comments below. If this is agreeable we anticipate starting this work in the next week.

**Comment by Anthchirp** [comment]
This work has now been completed.

* All tests pass on Python 2.7 and 3.6 in Jenkins and Travis
* We have achieved A+ rating on LGTM with 24 remaining warnings
* There are 25 remaining flake8 warnings
* Project closed 12 days ahead of schedule

Thanks to everyone involved.
